torch/nn/quantized/modules/linear.py

```diff
@@ -171,3 +171,54 @@ def from_float(mod):
         qlinear.out_scale = torch.tensor([act_qparams[0]])
         qlinear.out_zero_point = torch.tensor([act_qparams[1]])
         return qlinear
+
+
+class DynamicLinear(NNLinear):
+    r"""
+    A dynamic quantized linear module with quantized tensor as inputs
+    and outputs.
+    We adopt the same interface as `torch.nn.Linear`, please see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to `torch.nn.Linear`, attributes will be randomly initialized at
+        module creation time and will be overwritten later
+
+    Attributes:
+        weight: the non-learnable quantized weights of the
+                module which are of shape :math:`(\text{out\_features}, \text{in\_features})`.
+        bias:   the non-learnable bias of the module of shape :math:`(\text{out\_features})`.
+                If :attr:`bias` is ``True``, the values are initialized to zero.
+        scale: `scale` parameter of weight Quantized Tensor, type: double
+        zero_point: `zero_point` parameter for weight Quantized Tensor, type: long
+
+    Examples::
+
+        >>> m = nn.quantized.DynamicLinear(20, 30)
+        >>> input = torch.randn(128, 20)
+        >>> output = m(input)
+        >>> print(output.size())
+        torch.Size([128, 30])
+    """
+
+    __constants__ = ['bias', 'in_features', 'out_features', 'scale', 'zero_point']
+
+    def __init__(self, in_features, out_features, bias=True):
+        super(DynamicLinear, self).__init__(in_features, out_features, bias)
+        del self.weight
+        weight_fp32 = torch.Tensor(out_features, in_features).float()
+        self.register_buffer('weight', weight_fp32)
+        del self.bias
+        bias_fp32 = torch.Tensor(out_features).float()
+        self.register_buffer('bias', bias_fp32)
+
+        weight_prepack, col_offsets, self.scale, self.zero_point = torch.fbgemm_linear_quantize_weight(weight_fp32)
```

Comment:
since we're changing this code, can we also organize it better?

1) the output of this ideally should be a single struct packing all those args - can it be `PackedConvWeight` directly?
2) can this function just take quantized tensor and we use the same weight observer in python (so it's tweakable).  Ideally, we can just call WeightObserver here, get QTensor weight and then invoke fbgemm_linear_prepack which already exists

Reply:
We still want to keep quantization external to this function, similar to torch.nn.quantized.linear. This way, someone can quantize weights whichever way they want and still load the module with the quantized weights. I think (2) would become more like from_float(), which we will continue to support.

Reply:
I have updated this according to @dzhulgakov 's comment. See https://github.com/pytorch/pytorch/pull/23104 for the new op with ```PackedLinearWeight``` for (1), and https://github.com/pytorch/pytorch/pull/23128 for the updated version of torch.nn.quantized.Linear module with these updates.